### DISCUSSION TITLE
Getting NaN when training using FP16

### DISCUSSION THREAD

**Discussion by cmlakhan** [Lightning-AI/pytorch-lightning#12510] [code help: NLP / ASR / TTS]
I am trying to train a transformer model using FP16 precision, but the loss eventually goes to nan after around ~1000 steps.  I set the detect anomaly flag to True and I seem to be getting the following error below and am wondering if others have encountered this issue as well.  I have tried a variety of learning rates the current error comes from setting a low learning rate (1e-5).  I'm curious if others have encountered this issue.  I have had this same error with CNN models as well.  When I run the same model using FP32 precision it seems to be fine and no issues so is it something beyond a model issue?

![Screen Shot 2022-03-29 at 11 16 05 AM](https://user-images.githubusercontent.com/2146422/160645427-d76ec84a-7fd4-4b29-a740-a5a05a311dd7.png)

**Comment by akihironitta** [comment]
Hi @cmlakhan!

Generally speaking, training with fp16 is unstable due to its small range compared to fp32 and can lead to nan at some point, so instead, I'd also try bf16 whose range is wider than that of fp16.

```python
Trainer(..., precision="bf16")
```
